gcloud-java-datastore/src/main/java/com/google/gcloud/datastore/testing/LocalGcdHelper.java

```diff
@@ -179,49 +179,134 @@ private static Path executablePath(String cmd) {
+        // ignore
+      }
+    }
+
+    public static ProcessStreamReader start(InputStream inputStream) {
+      ProcessStreamReader thread = new ProcessStreamReader(inputStream);
+      thread.start();
+      return thread;
+    }
+  }
+
+  private static class ProcessErrorStreamReader extends Thread {
+    private static final int LOG_LENGTH_LIMIT = 50000;
+    private static final String GCD_LOGGING_CLASS =
+        "com.google.apphosting.client.serviceapp.BaseApiServlet";
+
+    private final BufferedReader errorReader;
+    private String currentLog = null;
+    private Level currentLogLevel = null;
+
+    ProcessErrorStreamReader(InputStream errorStream, String blockUntil) throws IOException {
+      super("Local GCD ErrorStream reader");
+      setDaemon(true);
+      errorReader = new BufferedReader(new InputStreamReader(errorStream));
       if (!Strings.isNullOrEmpty(blockUntil)) {
         String line;
         do {
-          line = reader.readLine();
+          line = errorReader.readLine();
         } while (line != null && !line.contains(blockUntil));
       }
     }
 
-    void terminate() throws InterruptedException, IOException {
-      process.destroy();
-      process.waitFor();
-      reader.close();
+    void terminate() throws IOException {
+      writeLog(currentLogLevel, currentLog);
+      errorReader.close();
     }
 
     @Override
     public void run() {
       try {
-        while (reader.readLine() != null) {
-          // consume line
+        boolean errorReaderDone = false;
+        String previousLine = "";
+        String currentLine = "";
+        while (!errorReaderDone) {
+          previousLine = currentLine;
+          currentLine = errorReader.readLine();
+          if (currentLine == null) {
+            errorReaderDone = true;
+          } else {
+            processLogLine(previousLine, currentLine);
+          }
         }
       } catch (IOException e) {
         // ignore
       }
     }
 
-    public static ProcessStreamReader start(Process process, String blockUntil) throws IOException {
-      ProcessStreamReader thread = new ProcessStreamReader(process, blockUntil);
+    private void processLogLine(String previousLine, String currentLine) {
+      // Each gcd log is two lines with the following format:
+      //     [Date] [Time] [GCD_LOGGING_CLASS] [method]
+      //     [LEVEL]: error message
+      // Exceptions and stack traces are included in gcd error stream, separated by a newline
+      Level nextLogLevel = getLevel(currentLine);
+      if (previousLine.contains(GCD_LOGGING_CLASS) && nextLogLevel != null) {
+        writeLog(currentLogLevel, currentLog);
+        if (currentLine.startsWith("SEVERE: ")) {
+          // don't show duplicate error messages from gcd.sh (see issue #258)
+          currentLog = null;
+          currentLogLevel = null;
+        } else {
+          currentLog = "GCD" + currentLine.split(":", 2)[1] + System.getProperty("line.separator");
```

Comment:
The `.split(":", 2)` splits the string into two pieces at the first occurence of ":" (confirmed with test).  Since we know the level is included in this particular line, I think we're okay.